### ISSUE TITLE
Free surface hydrodynamics

### ISSUE THREAD

**Issue by chapulina** [osrf/buoy_sim#21]
Gazebo's [Hydrodynamics](https://ignitionrobotics.org/api/gazebo/6.9/classignition_1_1gazebo_1_1systems_1_1Hydrodynamics.html) plugin is currently well-suited for subsurface links. We need to either extend that plugin, or create a new one that implements hydrodynamic forces at the surface, taking into account wave exciting, radiation and diffraction forces.

- [ ] Free surface hydrodynamics PR https://github.com/osrf/buoy_sim/pull/103 , dependent on #81
- [ ] hamilton8415: Some clean-up needed, there are a bunch of tests but most are specific to libFreeSurfaceHydrodynamics and aren't CI tests
- [x] hamilton8415: Currently, a library called libFreeSurfaceHydrodynamics.so  is being built out of WaveBodyInteractoins/lib, but isn't being installed in a usable place, so gz sim doesn't see it at start up.  I have been building this independently and copying it to the correct place, but surely this can be done automatically.  This arrangement doesn't seem great, I have considered pulling this library out of the repo entirely and hosting it elsewhere, as these computations are independent of gazebo, and have some unused capability that makes testing more complete, etc.
- [ ] hamilton8415: Currently, this plugin adds a bit of viscous drag to stabilize the model, without that the buoy drifts off to the edge of the world in time.  This can be taken out once the viscous drag plugin is added, this is an example of a feature in libFreeSurfaceHydrodynamics that isn't strictly required if the viscous plugin is in place.

Future work:
- [ ] Parameterize wave field (wave height, period, and spectrum), currently in SDF. Current and wind potential future adds
- [ ] Wave rendering based on Rhys Mainwaring's work. On hold.
  - Wave-forcing concentrates on the non-linear buoyancy forces and the non-linear wave-exciting; general wave-radiation forces based on an impulse response function close to completion.

**Comment by hamilton8415** [comment]
I have gotten a start on implementing some classes that generate incident waves and compute the resulting forces on the buoy as a result of those waves, and the forces that result from the buoy moving.  So far this is contained in two non-Gazebo classes, and is set up such that a different incident wave-generator class could be used reasonably easily.  The work of Rhys Mainwaring looks promising too.   I took a look at his code and it seems he's done a careful job of computing the buoyancy force based upon the instantaneous position of the buoy and water surface.  This is a fancy way of doing things, beyond what is usually done in ocean engineering.  On the other hand, for the damping part of the fluid-body interaction, it appears there is more reliance on heuristics.  In particular the damping that ordinary comes from the bodies tendency to generate waves is lumped in with the quadratic damping due to the viscosity of the fluid.  The FS_Hydrodyamics class implements a more physics based approach to generating the wave radiation forces, and could possibly be used in place of the Heuristics Rhys is using.  Those physics rely on coefficients generated by another program (WAMIT) that characterize the wave-body interaction dynamics.  The WAMIT users manual has some good background:  https://www.wamit.com/manualv7.4/wamit_v74manual.html    My intention at this point is to develop a simple Gazebo plugin on top of this code for demonstration purposes anyway, and we can see if it's a better path to refine that or try to collaborate with Rhys. He's done a lot of good work there, his wave-generation and visualization stuff is great, and he's included some of the non-linear fluid damping terms that are likely to be important.    Work is in branch hamilton8415/fs_hydrodynamics  At the moment I have CMakeLists.txt symbolically linked to a local CMakeLists file for ease of development on these non-Gazebo components...

**Comment by mjcarroll** [comment]
@hamilton8415 to open PR off of his branch.

**Comment by hamilton8415** [comment]
As threatened, I broke out the wave-body interaction force computation into a library independent of the Gazebo simulator.  This is now here:  https://github.com/hamilton8415/FreeSurfaceHydrodynamics

At this time in the hamilton8415/fs_hydrodynamics branch, the CI script installs this library from Debians in a PPA I made for this, so the building/installing is straightforward.  Seems to be working fine, we will see what issues of version-itis ensue...

Today I ended up having a zoom-call with Rhys Mainwaring (not a holiday in England, but it is raining...).  He is returning his focus to his wave-simulation, and by breaking these computations out he may be able to incorporate some of this work into his work. He's doing a heck of a job and there's still a good chance for this project to benefit from his plugins (and associated wave-rendering), he totally understands we needed a basic solution without his material though, so all is good.  

For now, I am keeping hamilton8415/fs_hydrodynamics up to date with the humble-garden branch, so will be ready to merge this into main as soon as humble-garden is merged.
